### PR TITLE
Classification-style tasks and training options with multiple input signals.

### PR DESCRIPTION
**Pull request by nortondeclan**
Classification-style tasks facilitated with updated train() and predict() routines. Options to batch and/or discard inputs and states during training to save memory. Time-delayed feature vectors allowed. Example scripts and one test added.

MRS and time-shifted features are still functions to enable jit compilation, which is useful since the MRS feature function is called repeatedly during training.  ESNFeature object versions of these features are provided in comments.

### CODE REVIEW

**Comment by nortondeclan** [comment]
I still need to update the test_rescompy.py set of tests.

File: rescompy/rescompy.py [hunk 1]
```diff
@@ -209,65 +208,126 @@ class TrainResult:
...
-        feature_function: Callable,
-        weights:          np.ndarray,
-        transient_length: int,
-        jacobian:         Optional[np.ndarray] = None
+        states:            List[np.ndarray],
+        inputs:            List[np.ndarray],
+        target_outputs:    Union[List[None],  List[np.ndarray]],
+        transient_lengths: List[int],
+        accessible_drives: List[int],
+        feature_function:  Callable,
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:222]: Should it be

`feature_function:  Union[features.ESNFeature, Callable],`

File: rescompy/rescompy.py [hunk 2]
```diff
@@ -209,65 +208,126 @@ class TrainResult:
...
-        inputs:           np.ndarray,
-        target_outputs:   np.ndarray,
-        feature_function: Callable,
-        weights:          np.ndarray,
-        transient_length: int,
-        jacobian:         Optional[np.ndarray] = None
+        states:            List[np.ndarray],
+        inputs:            List[np.ndarray],
+        target_outputs:    Union[List[None],  List[np.ndarray]],
+        transient_lengths: List[int],
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:220]: Does this mean a list of [None, None, None...?]  Why not `Union[None, List[np.ndarray]]` ?

File: rescompy/rescompy.py [hunk 3]
```diff
@@ -209,65 +208,126 @@ class TrainResult:
...
-        self.transient_length = transient_length
         self.jacobian = jacobian
         
+		# Check if the feature function requires a lookback period to include
+        # time-delayed states or inputs in feature vectors.
+        if hasattr(feature_function, 'lookback_length'):
+            self.lookback_length = feature_function.lookback_length
+        else: self.lookback_length = 0
+
+        self.listed_targets = target_outputs
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:282]: what are these listed_* properties?

File: rescompy/rescompy.py [hunk 4]
```diff
@@ -209,65 +208,126 @@ class TrainResult:
...
+                      "tasks employed different transient lengths."
+            logging.error(msg)
+        
     @property
     def features(self):
         """The features property.
         Computes the reservoir features from the states and inputs using the
         feature_function."""
-        return self.feature_function(self.states, self.inputs)
+        features = None
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:321]: maybe this should be done with a `functools.reduce()`?

File: rescompy/rescompy.py [hunk 5]
```diff
@@ -279,8 +339,9 @@ def reservoir_outputs(self):
     def rmse(self):
         """The root-mean-square error property.
         Computes the root-mean-square error as a function of time."""
-        return np.sqrt(np.mean(np.square(self.reservoir_outputs -
-                                         self.target_outputs), axis=1))
+        return np.sqrt(np.mean(
+			np.square(self.reservoir_outputs -
+			 self.target_outputs[self.lookback_length:]), axis=1))
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:345]: what is `lookback_length`?  is it zero usually?

File: rescompy/rescompy.py [hunk 6]
```diff
@@ -411,7 +521,12 @@ def nrmse(self):
         error will be raised.
         Is only defined when target_outputs are provided."""
         
-        if self.target_outputs is not None:
+        if self.predict_length == 0:
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:524]: should these cause errors or just return nan?

File: rescompy/rescompy.py [hunk 7]
```diff
@@ -452,37 +571,43 @@ def valid_length(
...
+                    msg = "Root-square-error does not exceed " \
                       f" {threshold}; true valid_length may be " \
                       "greater than reported."
-            logging.warning(msg)
+                logging.warning(msg)
+				
+            return valid_length
+
+        else:
+            msg = "valid_length is not defined for single-target mapping tasks" \
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:606]: should this be an error?

File: rescompy/rescompy.py [hunk 8]
```diff
@@ -740,48 +860,60 @@ def load(
...
         return _get_states_driven(inputs, states,
                                          self.A.data, self.A.indices,
                                          self.A.indptr, self.A.shape,
                                          self.B, self.C,
                                          self.leaking_rate)
-    
+
     @validate_arguments(config=dict(arbitrary_types_allowed=True))
-    def train(
+    def train_saving_memory(
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:879]: why is this renamed?

File: rescompy/rescompy.py [hunk 9]
```diff
@@ -815,98 +956,518 @@ def train(
...
+        target_outputs:    Union[np.ndarray, List[np.ndarray], None] = None,
+        initial_state:     Optional[np.ndarray]                      = None,
+        dg_du:             Optional[np.ndarray]                      = None,
+        feature_function:  Union[Callable, features.ESNFeature]      = features.StatesOnly(),
+        regression:        Optional[Callable]                        = regressions.batched_ridge(),
+        batch_size:        int                                       = 10,
+        accessible_drives: Union[int, List[int], str]                = "all",
+        ) -> TrainResult:
+        """The training method.
+
```
Comment (andrew-pomerance) [inline rescompy/rescompy.py:1097]: what is the difference between train and train batched?  maybe add some text here to the comment

File: rescompy/features/features.py [hunk 10]
```diff
@@ -200,9 +203,448 @@ def __call__(self, r, u):
...
+        self.compiled = numba.njit(lambda r, u: np.copy(r.reshape((-1, r.shape[-1]))[-1].reshape(1, -1)))
 
+    @staticmethod	
+    def feature_size(esn_size: int, signal_length: int):
+        return esn_size
+
+    @staticmethod
     def jacobian(r: np.ndarray, u: np.ndarray, dr_du: np.ndarray):
         raise NotImplementedError()
+
```
[the remark this answers is not in the document]
Comment (nortondeclan) [reply rescompy/features/features.py:242]: I left the MRS and lookback-related features as functions rather than classes so that they can be jit-compiled. Since they can't be written as lambda functions, right now I don't see a way to give a compiled option within a class definition. The jit-compilation is useful for the MRS feature function since it is called thousands (or tens of thousands) of times in the training stage.

That said, ESNFeature class versions of the MRS and lookback-related features are commented out at the bottom of the script. I could make them the default and just use my own jit-compiled version when I want to generate results.

File: rescompy/rescompy.py (hunk 1 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:222]: Yes, I'll change that.

File: rescompy/rescompy.py (hunk 5 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:345]: lookback_length is the greater of inputs_lookback_length and states_lookback_length i.e., the total lookback_length required to form feature vectors. It is zero whenever the features do not include time-delayed states or time-delayed inputs.

File: rescompy/rescompy.py (hunk 8 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:879]: I don't think it's renamed. It is a new function. I think it just appears to git that it has been renamed because train_saving_memory() appears earlier in the code than does train().

File: rescompy/rescompy.py (hunk 3 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:282]: self.listed_states stores the reservoir states from each training task as a separate entry in a list of arrays (one list entry/array per training task/input signal). self.states concatenates all the entries of self.listed_states in a single array, as in the current rescompy. The same is true for the other listed_* properties.

File: rescompy/rescompy.py (hunk 6 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:524]: I thought the error message might be useful so that the user knows that they are trying to access something that is not defined. Maybe returning nan but also printing/logging a warning message would be better.

File: rescompy/rescompy.py (hunk 7 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:606]: Maybe a warning message and returning nan would be better?

File: rescompy/rescompy.py (hunk 9 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:1097]: The batch training method is for use with regression functions that accept the regression matrices SS_T and VS_T as arguments. These matrices are incremented every batch_size training signals to save memory.

The "standard" train routine is for use with general regression functions that accept a combination of the inputs (u), features (s), targets (v), feature_function (g), or Jacobian (dg_du) as argument. The features, inputs, and targets over the entire training period are stored until training has finished.

Now that the signature of the regression function is checked before calling it, I may be able to merge them into the same function.However, I'm not sure how easily that could be done without sacrificing some of the memory that is saved in the current configuration.

Comment (nortondeclan) [reply rescompy/rescompy.py:1097]: I've added some text to the comments.

File: rescompy/features/features.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrew-pomerance) [reply rescompy/features/features.py:242]: gotcha.  you could do something like:

```
def __init__(self):
  def inner(args):
      # do something
  self.compiled = numba.njit(inner)
```

File: rescompy/rescompy.py (hunk 2 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:220]: I think I can remove the List[None]/None option altogether actually. I think that line is a holdover from a previous version of the code.

File: rescompy/features/features.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nortondeclan) [reply rescompy/features/features.py:242]: Works like a charm! Thanks. Added it to the polynomial ESNFeature too.

File: rescompy/rescompy.py (hunk 4 above)
Comment (nortondeclan) [reply rescompy/rescompy.py:321]: Done

File: rescompy/rescompy.py [hunk 11]
```diff
@@ -1021,14 +1090,14 @@ def train(
...
             utils.check_shape(inputs[task_ind].shape,
-                          (None, self.input_dimension), msg1)            
-
-        num_samples = len(inputs)
-        batch_size = min(num_samples, batch_size)
+                          (None, self.input_dimension), msg1)
+		
+        # If the feature_function has a compiled option, use it. Otherwise
+		# use the non-compiled version.
+        if hasattr(feature_function, 'compiled'):
```
Comment (nortondeclan) [inline rescompy/rescompy.py:1097]: Would this be better as:

if hasattr(feature_function, 'compiled') and \
			issubclass(type(feature_function), features.SingleFeature):
    feature_function_jit = feature_function.compiled
else: feature_function_jit = feature_function

Jit-compilation of the feature function gave me about 10% quicker run-times in a few MRS-style tasks that I tried.  For tasks with fewer sample input signals compilation slowed training slightly.  My guess is that most classification-style tasks will require enough sample inputs for training to make the compilation worth it, but most time-series prediction tasks will use fewer sample inputs. So maybe the default should be to jit-compile only when using a features.SingleFeature feature function.

**Review by andrew-pomerance — APPROVED** [review]
